File: 19_数学/计算几何/直线/149. 直线上最多的点数.java
import java.util.HashMap;
import java.util.Map;

class Solution {
  public int maxPoints(int[][] ps) {
    int n = ps.length;
    int ans = 1;
    for (int i = 0; i < n; i++) {
      Map<String, Integer> map = new HashMap<>();
      // 由当前点 i 发出的直线所经过的最多点数量
      int max = 0;
      for (int j = i + 1; j < n; j++) {
        int x1 = ps[i][0], y1 = ps[i][1], x2 = ps[j][0], y2 = ps[j][1];
        int a = x1 - x2, b = y1 - y2;
        int k = gcd(a, b);
        String key = (a / k) + "_" + (b / k);
        map.put(key, map.getOrDefault(key, 0) + 1);
        max = Math.max(max, map.get(key));
      }
      ans = Math.max(ans, max + 1);
    }
    return ans;
  }

  static int gcd(int a, int b) {
    return b == 0 ? a : gcd(b, a % b);
  }

  public static void main(String[] args) {
    int[][] ps = { { 1, 1 }, { 3, 2 }, { 5, 3 }, { 4, 1 }, { 2, 3 }, { 1, 4 } };
    System.out.println(new Solution().maxPoints(ps));
    System.out.println(gcd(-2, -1));
  }
}

File: 19_数学/计算几何/直线/practice/149. 直线上最多的点数.py
from collections import defaultdict
from decimal import Decimal
from fractions import Fraction
from math import gcd
from typing import List, Tuple

# !求最多有多少个点在同一条直线上。
INF = int(1e18)


def calSlope1(x1: int, y1: int, x2: int, y2: int) -> Tuple[int, int]:
    """直线斜率"""
    if x2 == x1:
        return (INF, INF)
    gcd_ = gcd(x2 - x1, y2 - y1)
    a, b = (y2 - y1) // gcd_, (x2 - x1) // gcd_
    if a == 0:
        return (0, b if b > 0 else -b)
    elif a < 0:
        return (-a, -b)
    else:
        return (a, b)


class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        """两点确定一条直线"""
        n, res = len(points), 1
        for i in range(n):
            x1, y1 = points[i]
            slopeCounter = defaultdict(int)  # 点i加斜率唯一确定一条直线
            for j in range(i + 1, n):
                x2, y2 = points[j]
                slope = calSlope1(x1, y1, x2, y2)
                print(slope)
                slopeCounter[slope] += 1

            for count in slopeCounter.values():
                res = max(res, count + 1)

        return res


if __name__ == "__main__":
    print(Solution().maxPoints([[1, 1], [2, 2], [3, 3]]))
    print(Solution().maxPoints([[2, 3], [3, 3], [-5, 3]]))
